### PR TITLE
Automatically enforce coding guidelines

### PR DESCRIPTION
**Pull request by Ingrater**
Use clang-tidy to enforce our coding guidelines.

This PR currently consits of these parts:

* The required changes to make ezEngine build with clang on windows.
* A textual description of our coding guidelines.
* The changes done by clang-tidy to enforce our coding guidelines.

Missing and coming in the future:
* Source code and build instructions for the clang-tidy modification / extension.

Please comment if you do not like any of the changes done by clang-tidy or the coding guidelines in general.

### CODE REVIEW

**Comment by jankrassnigg** [comment]
> The required changes to make ezEngine build with clang on windows

Without yet looking at the PR itself, I would do this part as a separate PR first.

**Comment by Ingrater** [comment]
> Without yet looking at the PR itself, I would do this part as a separate PR first.

I can do that. Still interrested in the opinions on the rest of the changes.

File: Code/Editor/EditorEngineProcessFramework/EngineProcess/EngineProcessViewContext.h [hunk 1]
```diff
@@ -33,15 +33,15 @@ class EZ_EDITORENGINEPROCESSFRAMEWORK_DLL ezEditorProcessViewWindow : public ezW
...
+  virtual ezWindowHandle GetNativeWindowHandle() const override { return m_pWnd; }
   virtual void ProcessWindowMessages() override {}
   virtual bool IsFullscreenWindow(bool bOnlyProperFullscreenMode = false) const override { return false; }
 
   ezUInt16 m_uiWidth;
   ezUInt16 m_uiHeight;
 
 private:
-  ezWindowHandle m_hWnd;
+  ezWindowHandle m_pWnd;
```
Comment (jankrassnigg) [inline Code/Editor/EditorEngineProcessFramework/EngineProcess/EngineProcessViewContext.h]: HWND should stay hWnd

File: Code/Editor/EditorFramework/Assets/AssetBrowserFilter.moc.h [hunk 2]
```diff
@@ -46,11 +46,11 @@ class EZ_EDITORFRAMEWORK_DLL ezQtAssetBrowserFilter : public ezQtAssetFilter
   bool m_bShowItemsInSubFolders = true;
   bool m_bShowItemsInHiddenFolders = false;
   bool m_bSortByRecentUse = false;
-  mutable ezStringBuilder m_sTemp; // stored here to reduce unnecessary allocations
+  mutable ezStringBuilder m_Temp; // stored here to reduce unnecessary allocations
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/Assets/AssetBrowserFilter.moc.h]: I would have expected ezStringBuilder to have an s prefix

File: Code/Editor/EditorFramework/Assets/AssetProcessor.h [hunk 3]
```diff
@@ -64,7 +64,7 @@ class EZ_EDITORFRAMEWORK_DLL ezAssetProcessor
 private:
   mutable ezMutex m_ProcessorMutex;
   ezAssetProcessorLog m_CuratorLog;
-  ezAtomicInteger32 m_bRunProcessTask;
+  ezAtomicInteger32 m_RunProcessTask;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/Assets/AssetProcessor.h]: For a long time we didn't have atomic bools, so this is probably why we use an int as a bool here.
I think atomics should also use "i" "u" and "b".

File: Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h [hunk 4]
```diff
@@ -28,6 +28,6 @@ private Q_SLOTS:
 
   ezUInt8 m_uiCurDesc = 0;
   ezWindowCreationDesc m_Descs[2];
-  bool m_bOverrideProjectDefault[2];
+  bool m_OverrideProjectDefault[2];
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h]: I guess you are ignoring arrays ?

File: Code/Editor/EditorEngineProcessFramework/EngineProcess/EngineProcessViewContext.h [hunk 5]
```diff
@@ -33,15 +33,15 @@ class EZ_EDITORENGINEPROCESSFRAMEWORK_DLL ezEditorProcessViewWindow : public ezW
 
   // Inherited via ezWindowBase
   virtual ezSizeU32 GetClientAreaSize() const override { return ezSizeU32(m_uiWidth, m_uiHeight); }
-  virtual ezWindowHandle GetNativeWindowHandle() const override { return m_hWnd; }
+  virtual ezWindowHandle GetNativeWindowHandle() const override { return m_pWnd; }
```
Comment (C-Core) [inline Code/Editor/EditorEngineProcessFramework/EngineProcess/EngineProcessViewContext.h]: ezWindowHandle should be a handle type, thus the h prefix should stay

File: Code/Editor/EditorFramework/DocumentWindow/EngineViewWidget.moc.h [hunk 6]
```diff
@@ -147,11 +147,11 @@ private Q_SLOTS:
   float m_fCameraLerp;
   float m_fCameraStartFovOrDim;
   float m_fCameraTargetFovOrDim;
-  ezVec3 m_vCameraStartPosition;
-  ezVec3 m_vCameraTargetPosition;
-  ezVec3 m_vCameraStartDirection;
-  ezVec3 m_vCameraTargetDirection;
-  ezVec3 m_vCameraUp;
+  ezVec3 m_CameraStartPosition;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/DocumentWindow/EngineViewWidget.moc.h]: Yeah... we should decide whether we use a prefix for vectors or not.

File: Code/Editor/EditorFramework/Assets/AssetBrowserFilter.moc.h (hunk 2 above)
Comment (C-Core) [inline Code/Editor/EditorFramework/Assets/AssetBrowserFilter.moc.h]: ezStringBuilder is also a string type

File: Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h [hunk 7]
```diff
@@ -265,21 +265,21 @@ private Q_SLOTS:
...
   void SaveRecentFiles();
   void LoadRecentFiles();
 
-  ezRecentFilesList s_RecentProjects;
-  ezRecentFilesList s_RecentDocuments;
+  ezRecentFilesList m_RecentProjects;
+  ezRecentFilesList m_RecentDocuments;
 
-  int m_iArgc = 0;
+  int m_Argc = 0;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h]: We could probably enforce that int = ezInt32 ?

File: Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h [hunk 8]
```diff
@@ -265,21 +265,21 @@ private Q_SLOTS:
...
   QApplication* m_pQtApplication = nullptr;
   ezLongOpControllerManager m_LongOpControllerManager;
-  ezEditorEngineProcessConnection* s_pEngineViewProcess;
+  ezEditorEngineProcessConnection* m_pEngineViewProcess;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h:282]: Can we enforce that pointers are nullptr initialized ? It's the one thing that IMO should never be left uninitialized.

File: CodingGuidelines.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Coding Guidelines
+
+The coding guidelines in ezEngine are enforced through clang-tidy. You can either run clang-tidy locally or use the automated CI process that runs in every PR into ezEngine. CI will provide a git patch with all suggested changes which you can apply locally. After applying the suggested changes please make sure everything still compiles. The fixes done by clang-tidy are not garantueed to work in all cases.
+
+## Type dependant prefixes
```
Comment (C-Core) [inline CodingGuidelines.md]: typo: should be dependent

File: Code/Editor/EditorFramework/GUI/RawDocumentTreeWidget.moc.h [hunk 10]
```diff
@@ -38,8 +40,8 @@ private Q_SLOTS:
   void SelectionEventHandler(const ezSelectionManagerEvent& e);
 
 private:
-  std::unique_ptr<ezQtDocumentTreeModel> m_pModel;
-  std::unique_ptr<ezQtTreeSearchFilterModel> m_pFilterModel;
+  std::unique_ptr<ezQtDocumentTreeModel> m_Model;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/GUI/RawDocumentTreeWidget.moc.h]: std::unique_ptr and shared_ptr should also be treated as a pointer

File: CodingGuidelines.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+## Members of structs and Classes
+
+### Non-Static
+
+If the member is public, no rules apply.
+
+For private and protected members, the following rules apply:
+ * All members must start with 'm_' (this comes before the type dependant prefix)
+ * The name of the static member must be in PascalCase: `s_MyMember`
```
Comment (C-Core) [inline CodingGuidelines.md]: this is the non static member section, thus this sentence and example is weird

**Comment by jankrassnigg** [comment]
Can/should we apply these rules also to function argument names?
There we could also think about enforcing stuff like "out" or "inout" for non-const references.

File: Code/Editor/EditorFramework/Assets/AssetWatcher.h [hunk 12]
```diff
@@ -51,7 +51,7 @@ class EZ_EDITORFRAMEWORK_DLL ezAssetWatcher
   ezHybridArray<ezDirectoryWatcher*, 6> m_Watchers;
   ezSharedPtr<ezTask> m_pWatcherTask;
   ezTaskGroupID m_WatcherGroup;
-  ezAtomicBool m_bShutdown = false;
+  ezAtomicBool m_Shutdown = false;
```
Comment (C-Core) [inline Code/Editor/EditorFramework/Assets/AssetWatcher.h]: Should we do the same prefixes for atomics as for regular vars? I have no preference tbh.

File: Code/Editor/EditorFramework/PropertyGrid/ExposedBoneWidget.moc.h [hunk 13]
```diff
@@ -27,5 +27,5 @@ private Q_SLOTS:
 
   bool m_bTemporaryCommand = false;
   QHBoxLayout* m_pLayout = nullptr;
-  ezQtDoubleSpinBox* m_pRotWidget[3];
+  ezQtDoubleSpinBox* m_RotWidget[3];
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/PropertyGrid/ExposedBoneWidget.moc.h]: arrays again

File: Code/Editor/EditorFramework/PropertyGrid/GameObjectReferencePropertyWidget.moc.h [hunk 14]
```diff
@@ -35,10 +37,10 @@ protected slots:
   virtual void showEvent(QShowEvent* event) override;
 
 protected:
-  QPalette m_pal;
+  QPalette m_Pal;
   QHBoxLayout* m_pLayout = nullptr;
   QLabel* m_pWidget = nullptr;
-  QString m_sInternalValue;
+  QString m_InternalValue;
```
Comment (jankrassnigg) [inline Code/Editor/EditorFramework/PropertyGrid/GameObjectReferencePropertyWidget.moc.h]: QString

File: Code/EditorPlugins/Assets/EditorPluginAssets/ImageDataAsset/ImageDataAssetWindow.moc.h [hunk 15]
```diff
@@ -20,10 +24,10 @@ class ezQtImageDataAssetDocumentWindow : public ezQtDocumentWindow
 
 private:
   void ImageDataAssetEventHandler(const ezImageDataAssetEvent& e);
-  ezEvent<const ezImageDataAssetEvent&>::Unsubscriber m_eventUnsubscriper;
+  ezEvent<const ezImageDataAssetEvent&>::Unsubscriber m_EventUnsubscriper;
 
   void UpdatePreview();
 
-  QPointer<ezQtImageWidget> m_pImageWidget;
+  QPointer<ezQtImageWidget> m_ImageWidget;
```
Comment (jankrassnigg) [inline Code/EditorPlugins/Assets/EditorPluginAssets/ImageDataAsset/ImageDataAssetWindow.moc.h]: not sure about this one

File: Code/Editor/EditorFramework/Assets/AssetProcessor.h (hunk 3 above)
Comment (Ingrater) [reply Code/Editor/EditorFramework/Assets/AssetProcessor.h]: What do you mean by 'i, 'u' and 'b'? There is a Single atomic type. So it needs a Single prefix?

File: Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h (hunk 4 above)
Comment (Ingrater) [reply Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h]: Well, the rule is 'if it is a bool use the b prefix' not 'if it is a bool array'. If we do this for bool, will we do it for all other prefixes as well?

File: Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h (hunk 7 above)
Comment (Ingrater) [reply Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h]: What exactly do you mean by that? Enforce where?

File: Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h (hunk 8 above)
Comment (Ingrater) [reply Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h:282]: Yes, we csn do that with clang-tidy.

File: Code/EditorPlugins/Scene/EnginePluginScene/SceneContext/SceneContext.h [hunk 16]
```diff
@@ -120,7 +120,7 @@ class EZ_ENGINEPLUGINSCENE_DLL ezSceneContext : public ezEngineProcessDocumentCo
   ezDeque<ezGameObjectHandle> m_Selection;
   ezDeque<ezGameObjectHandle> m_SelectionWithChildren;
   ezSet<ezGameObjectHandle> m_SelectionWithChildrenSet;
-  ezGameObjectHandle m_hAmbientLight[3];
```
Comment (jankrassnigg) [inline Code/EditorPlugins/Scene/EnginePluginScene/SceneContext/SceneContext.h]: more arrays

File: Code/Engine/Core/System/Window.h [hunk 17]
```diff
@@ -331,7 +331,7 @@ class EZ_CORE_DLL ezWindow : public ezWindowBase
 
   ezUniquePtr<ezStandardInputDevice> m_pInputDevice;
 
-  mutable ezWindowInternalHandle m_WindowHandle = ezWindowInternalHandle();
+  mutable ezWindowInternalHandle m_pWindowHandle = ezWindowInternalHandle();
```
Comment (jankrassnigg) [inline Code/Engine/Core/System/Window.h]: handle

File: Code/Engine/Core/World/Component.h [hunk 18]
```diff
@@ -230,7 +230,7 @@ class EZ_CORE_DLL ezComponent : public ezReflectedClass
   ezComponentManagerBase* m_pManager = nullptr;
   ezGameObject* m_pOwner = nullptr;
 
-  static ezWorldModuleTypeId TYPE_ID;
+  static ezWorldModuleTypeId s_TYPE_ID;
```
Comment (jankrassnigg) [inline Code/Engine/Core/World/Component.h]: if it's already all caps, should we treat it differently ?

File: Code/Engine/Foundation/CodeUtils/Expression/Implementation/ExpressionParser_inl.h [hunk 19]
```diff
@@ -31,7 +31,7 @@ inline ezResult ezExpressionParser::Expect(ezTokenType::Enum Type, const ezToken
   {
     const ezUInt32 uiErrorToken = ezMath::Min(m_TokenStream.GetCount() - 1, m_uiCurrentToken);
     auto pToken = m_TokenStream[uiErrorToken];
-    ReportError(pToken, ezFmt("Syntax error, expected token type {} but got {}", ezTokenType::EnumNames[Type], ezTokenType::EnumNames[pToken->m_iType]));
+    ReportError(pToken, ezFmt("Syntax error, expected token type {} but got {}", ezTokenType::s_EnumNames[Type], ezTokenType::s_EnumNames[pToken->m_iType]));
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/CodeUtils/Expression/Implementation/ExpressionParser_inl.h]: don't like this change, though maybe the type should be something else (constexpr or so)

File: Code/Engine/Foundation/CodeUtils/MathExpression.h [hunk 20]
```diff
@@ -46,9 +46,9 @@ class EZ_FOUNDATION_DLL ezMathExpression
   float Evaluate(ezArrayPtr<Input> inputs = ezArrayPtr<Input>()); // [tested]
 
 private:
-  ezHashedString m_OriginalExpression;
+  ezHashedString m_sOriginalExpression;
   bool m_bIsValid = false;
 
   ezExpressionByteCode m_ByteCode;
-  ezExpressionVM m_VM;
+  ezExpressionVM m_M;
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/CodeUtils/MathExpression.h]: WTF ?

File: Code/Engine/Foundation/Communication/Implementation/Win/MessageLoop_win.h [hunk 21]
```diff
@@ -41,8 +41,8 @@ class EZ_FOUNDATION_DLL ezMessageLoop_win : public ezMessageLoop
 
 private:
   ezDynamicArray<IOItem> m_CompletedIO;
-  LONG m_haveWork = 0;
-  HANDLE m_Port = INVALID_HANDLE_VALUE;
+  LONG m_HaveWork = 0;
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Communication/Implementation/Win/MessageLoop_win.h]: are LONGs ints and are HANDLEs handles ?

File: Code/Editor/EditorFramework/Assets/AssetProcessor.h (hunk 3 above)
Comment (jankrassnigg) [reply Code/Editor/EditorFramework/Assets/AssetProcessor.h]: We have ezAtomicInt(32/64) and ezAtomicBool now. So should use i and b.

File: Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h (hunk 4 above)
Comment (jankrassnigg) [reply Code/Editor/EditorFramework/Dialogs/WindowCfgDlg.moc.h]: yes

File: Code/Editor/EditorFramework/DocumentWindow/EngineViewWidget.moc.h (hunk 6 above)
Comment (jankrassnigg) [reply Code/Editor/EditorFramework/DocumentWindow/EngineViewWidget.moc.h]: Also quaternions yes/no/maybe

File: Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h (hunk 7 above)
Comment (jankrassnigg) [reply Code/Editor/EditorFramework/EditorApp/EditorApp.moc.h]: I mean automatically change "int" to "ezInt32" etc

File: CodingGuidelines.md [hunk 22]
```diff
@@ -0,0 +1,34 @@
+# Coding Guidelines
+
+The coding guidelines in ezEngine are enforced through clang-tidy. You can either run clang-tidy locally or use the automated CI process that runs in every PR into ezEngine. CI will provide a git patch with all suggested changes which you can apply locally. After applying the suggested changes please make sure everything still compiles. The fixes done by clang-tidy are not garantueed to work in all cases.
+
+## Type dependant prefixes 
+
+These type dependant prefix is currently only mandatory for member variables. Function / Method parameter names and variables in Function / Method bodies can be named to the programmers liking.
```
Comment (jankrassnigg) [inline CodingGuidelines.md]: These type dependant prefixes are currently ...

File: CodingGuidelines.md [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+## Type dependant prefixes 
+
+These type dependant prefix is currently only mandatory for member variables. Function / Method parameter names and variables in Function / Method bodies can be named to the programmers liking.
+
+* If the variable / member is a `ezInt8`, `ezInt16`, `ezInt32`, `ezInt64` or `ptrdiff_t` the prefix is 'i': `ezInt32 iMyVar;`
+ * If the member is a `ezUint8`, `ezUInt16,` `ezUInt32`, `ezUInt64` or `size_t` the prefix is 'ui': `ezUInt32 uiMyVar;`
+ * If the member is a `float` or `double` the prefix is 'f': `float fMyVar;`
+ * If the member is a `bool` the prefix is 'b': `bool bMyVar;`
+ * If the member is a handle, the prefix is 'h': `ezSpatialDataHandle hMyVar;`
+ * If the member is a raw pointer, `ezSharedPtr` or `ezUniquePtr` the prefix is 'p': `ezUInt32* MyVar;`
```
Comment (jankrassnigg) [inline CodingGuidelines.md]: pMyVar

File: CodingGuidelines.md [hunk 24]
```diff
@@ -0,0 +1,34 @@
+# Coding Guidelines
+
+The coding guidelines in ezEngine are enforced through clang-tidy. You can either run clang-tidy locally or use the automated CI process that runs in every PR into ezEngine. CI will provide a git patch with all suggested changes which you can apply locally. After applying the suggested changes please make sure everything still compiles. The fixes done by clang-tidy are not garantueed to work in all cases.
+
+## Type dependant prefixes 
+
+These type dependant prefix is currently only mandatory for member variables. Function / Method parameter names and variables in Function / Method bodies can be named to the programmers liking.
+
+* If the variable / member is a `ezInt8`, `ezInt16`, `ezInt32`, `ezInt64` or `ptrdiff_t` the prefix is 'i': `ezInt32 iMyVar;`
```
Comment (jankrassnigg) [inline CodingGuidelines.md]: The indentation of the bullet points is inconsistent

File: CodingGuidelines.md (hunk 11 above)
Comment (jankrassnigg) [reply CodingGuidelines.md]: I think this is copy paste. Just remove the 'static' and change to m_MyMember

File: Code/Engine/Foundation/CodeUtils/Expression/Implementation/ExpressionParser_inl.h (hunk 19 above)
Comment (Ingrater) [reply Code/Engine/Foundation/CodeUtils/Expression/Implementation/ExpressionParser_inl.h]: This is a leftover change from an older version of my clang-tidy. The new version will no longer touch public statics.

File: Code/Engine/Foundation/CodeUtils/MathExpression.h (hunk 20 above)
Comment (Ingrater) [reply Code/Engine/Foundation/CodeUtils/MathExpression.h]: clang-tidy though here that 'V' is the prefix for 'M'. fixed.

**Comment by Ingrater** [comment]
> Can/should we apply these rules also to function argument names?
> There we could also think about enforcing stuff like "out" or "inout" for non-const references.

As a first steps I want to enforce rules on member variables only (static & non-static). This limits the scope to something reasonable. Function argument names will be done in a separate step.

File: .github/workflows/build-release-package.yml [hunk 25]
```diff
@@ -98,7 +98,7 @@ jobs:
 
       - name: Create 7z Archive
         run: |
-          repo\Data\Tools\Precompiled\7za.exe a ezEngine.Release.${{ steps.sdk-version.outputs.content }}.7z .\_package\* -mx9
+          repo\Data\Tools\Precompiled\7z.exe a ezEngine.Release.${{ steps.sdk-version.outputs.content }}.7z .\_package\* -mx9
```
Comment (jankrassnigg) [inline .github/workflows/build-release-package.yml:101]: I'm pretty sure that's not the file that we have checked in, no?

**Review by jankrassnigg — APPROVED** [review]

File: .github/workflows/build-release-package.yml (hunk 25 above)
Comment (Ingrater) [reply .github/workflows/build-release-package.yml:101]: 7za.exe does not Support unpacking Nsis installers. This is needed for unpacking llvm though. So I replaced the 7za.exe with 7z.exe which does Support unpacking Nsis

**Pull request merged**